ttcn3/types/types.go

```diff
@@ -399,5 +399,35 @@ func printExpr(e syntax.Expr) string {
 }
 
 func TypeOf(n syntax.Expr) Type {
+	//types to check:
+	//ValueLiteral, CompositeLiteral, BinaryExpr, UnaryExpr, Ident
+	switch n := n.(type) {
+	case *syntax.ValueLiteral:
+		switch n.Tok.Kind() {
+		case syntax.INT:
+			return Predefined["integer"]
+		case syntax.FLOAT, syntax.NAN:
+			return Predefined["float"]
+		case syntax.TRUE, syntax.FALSE:
+			return Predefined["boolean"]
+		case syntax.PASS, syntax.FAIL, syntax.INCONC, syntax.NONE, syntax.ERROR:
+			return Predefined["verdicttype"]
+		case syntax.STRING:
+			if strings.IndexFunc(n.Tok.String(), func(r rune) bool {
+				return r <= 32 || r >= 126
+			}) > -1 {
+				return Predefined["universal charstring"]
+			} else {
+				return Predefined["charstring"]
+			}
+		case syntax.BSTRING:
+			if strings.Contains(n.Tok.String(), "H") {
+				return Predefined["hexstring"]
+			} else if strings.Contains(n.Tok.String(), "O") {
+				return Predefined["octetstring"]
+			}
+			return Predefined["bitstring"]
```

Comment:
Checking the whole string for H or B is too much, also I don't think `bitstring` is the best default. I would use nil as default until we know better. How about something like:
```suggestion
			s := n.Tok.String()
			if len(s) == 0 {
				return nil
			}
			switch s[len(s)-1] {
			case 'H', 'h':
				return Predefined["hexstring"]
			case 'O','o':
				return Predefined["octetstring"]
			case 'B','b':
				return Predefined["bitstring"]
			}
			return nil
```